conmon-rs/server/src/streams.rs

```diff
@@ -75,52 +86,72 @@ impl Streams {
 
     fn read_loop(
         message_tx: Sender<Message>,
-        stop_rx: CrossbeamReceiver<()>,
-        stdout: Stream,
-        stderr: Stream,
+        stop_rx_stdout: broadcast::Receiver<()>,
+        stop_rx_stderr: broadcast::Receiver<()>,
+        stdout: RawFd,
+        stderr: RawFd,
     ) {
         debug!("Start reading from IO streams");
 
         let message_tx_stdout = message_tx.clone();
-        let stop_rx_clone = stop_rx.clone();
 
-        thread::spawn(move || {
-            Self::read_loop_single_stream(message_tx_stdout, stop_rx_clone, stdout)
+        task::spawn(async move {
+            Self::read_loop_single_stream(message_tx_stdout, stop_rx_stdout, stdout).await
+        });
+        task::spawn(async move {
+            Self::read_loop_single_stream(message_tx, stop_rx_stderr, stderr).await
         });
-        thread::spawn(move || Self::read_loop_single_stream(message_tx, stop_rx, stderr));
     }
 
-    fn read_loop_single_stream(
+    async fn read_loop_single_stream(
         message_tx: Sender<Message>,
-        stop_rx: CrossbeamReceiver<()>,
-        stream: Stream,
+        mut stop_rx: broadcast::Receiver<()>,
+        fd: RawFd,
     ) -> Result<()> {
-        trace!("Start reading from single stream: {:?}", stream);
+        trace!("Start reading from single fd: {:?}", fd);
 
         let message_tx_clone = message_tx.clone();
-        thread::spawn(move || {
-            let mut reader = BufReader::new(stream);
-            let mut buf = vec![0; 1024];
-            loop {
-                match reader.read(&mut buf) {
+        let (thread_shutdown_tx, thread_shutdown_rx) = oneshot::channel();
+        task::spawn(async move {
+            Self::run_buffer_loop(message_tx_clone, fd, thread_shutdown_rx).await
+        });
+
+        stop_rx
+            .recv()
+            .await
+            .context("unable to wait for stop channel")?;
+        debug!("Received IO stream stop signal");
+        thread_shutdown_tx
+            .send(())
+            .map_err(|_| format_err!("unable to send thread shutdown message"))?;
+        message_tx.send(Message::Done).context("send done message")
+    }
+
+    async fn run_buffer_loop(
+        message_tx: Sender<Message>,
+        fd: RawFd,
+        mut thread_shutdown_rx: oneshot::Receiver<()>,
+    ) {
+        let stream = unsafe { File::from_raw_fd(fd) };
```

Comment:
It looks odd to me to have this here - this value is being sent over a channel right?  I think we should maintain it as a `File` instance in the channel instead.  That will ensure that the fd is closed on panic/drop correctly.

Reply:
Hm, no this is just for making reading from the file descriptor easier.

Reply:
OK right, this isn't related to this PR, but it'd be better to do:

```
        let (stdout_fd_read, stdout_fd_write) =
            unistd::pipe2(OFlag::O_CLOEXEC).context("create stdout pipe")?;
        let (stdout_fd_read, stdout_fd_write) = unsafe {
           (File::from_raw_fd(stdout_fd_read), File::from_raw_fd(stdout_fd_write))
        }
```

and then pass them around as owned `File` instances.  

It's a bit surprising to me that the `nix` bindings for `pipe` give `RawFd` instances.  It looks like [the rustix version](https://docs.rs/rustix/latest/rustix/io/fn.pipe.html) is better here.

Reply:
Good? I’ll follow up on that when this PR got merged. 👍